File: Leetcode/Spiral Matrix.cc
const int mx[] = {1, 0, -1, 0};
const int my[] = {0, 1, 0, -1};
const int VISITED = 0x3f3f3f3f;

class Solution {
private:
    int n, m;
public:
    vector<int> spiralOrder(vector<vector<int> > &matrix) {
        vector<int> res;
        int dir = 0;
        n = matrix.size();
        if (n == 0) {
            return res;
        }
        m = matrix[0].size();
        int y = 0, x = 0;
        while (1) {
            res.push_back(matrix[y][x]);
            matrix[y][x] = VISITED;
            dir = next_dir(y, x, dir, matrix);
            y += my[dir];
            x += mx[dir];
            if (dir == -1) {
                break;
            }
        }
        return res;
    }
private:
    int next_dir(int y, int x, int dir, vector<vector<int> > &matrix) {
        for (int i = 0; i < 4; i++) {
            int ny = y + my[dir];
            int nx = x + mx[dir];
            if (ny < n && ny >= 0 && nx < m && nx >= 0 && matrix[ny][nx] != VISITED) {
                return dir;
            }
            dir = (dir + 1) % 4;
        }
        return -1;  
    }
};

File: Leetcode/Algorithm/golang/Spiral Matrix.go
// https://leetcode.com/problems/spiral-matrix/submissions/

const INF int = 0x3f3f3f3f

func spiralOrder(matrix [][]int) []int {
    var res []int
    
    n := len(matrix)
    if n == 0 {
        return res
    }
    m := len(matrix[0])
    
    y := 0
    x := 0
    dir := 0
    
    my := [4]int {0, 1, 0, -1}
    mx := [4]int {1, 0, -1, 0}
     
    for i := 0; i < n * m; i++ {
        res = append(res, matrix[y][x])
        matrix[y][x] = -INF
        for i < n * m - 1 {
            ny := y + my[dir]
            nx := x + mx[dir]

            if (ny < 0 || ny >= n || nx < 0 || nx >= m || matrix[ny][nx] == -INF) {
                dir = (dir + 1) % 4
            } else {
                y = ny
                x = nx
                break
            }
        }
    }
    return res
}
